dcap/ql/src/quote_enclave.rs

```diff
@@ -8,17 +8,167 @@
 use mc_sgx_core_types::TargetInfo;
 use mc_sgx_dcap_ql_types::PathKind;
-use mc_sgx_dcap_types::{Quote3Error, RequestPolicy};
+use mc_sgx_dcap_types::RequestPolicy;
 use mc_sgx_util::ResultInto;
-use std::{ffi::CString, os::unix::ffi::OsStrExt, path::Path};
+use once_cell::sync::Lazy;
+use std::{ffi::CString, os::unix::ffi::OsStrExt, path::Path, sync::Mutex};
+
+/// A convenience type alias for a `Result` which contains an [`Error`].
+pub type Result<T> = CoreResult<T, Error>;
+
+/// Initialization of the paths for the quoting enclaves and quote provider
+/// library
+///
+/// This should only be called once during process start up utilizing
+/// [`PathInitializer::with_paths()`] or [`PathInitializer::try_default()`].
+/// If a consumer of this crate does not explicitly initialize the paths, then
+/// they will be defaulted on the first call to an SGX function that needs the
+/// paths set.
+#[derive(Debug)]
+pub struct PathInitializer;
+
+static PATH_INITIALIZER: Lazy<Mutex<Option<PathInitializer>>> = Lazy::new(|| Mutex::new(None));
+
+impl PathInitializer {
+    /// Try to initialize the paths to the default for the system
+    ///
+    /// Currently the defaults assume the default DCAP install on an Ubuntu
+    /// machine.
+    ///
+    /// # Errors
+    /// * [`Error::PathsInitialized`] if the paths have been previously
+    ///   initialized.
+    /// * [`Error::Sgx`] if any of the default paths don't exist on the system.
+    pub fn try_default() -> Result<()> {
+        Self::with_paths(
+            "/usr/lib/x86_64-linux-gnu/libsgx_qe3.signed.so.1",
+            "/usr/lib/x86_64-linux-gnu/libsgx_pce.signed.so.1",
+            None,
+            "/usr/lib/x86_64-linux-gnu/libsgx_id_enclave.signed.so.1",
+        )
+    }
+
+    /// Initialize the DCAP quoting library paths with provided values
+    ///
+    /// # Arguments
+    /// * `quoting_enclave` - The full file path to the quoting enclave
+    /// * `provisioning_certificate_enclave` - The full file path to the
+    ///   provisioning certificate enclave
+    /// * `quote_provider_library` - The full file path to the quote provider
+    ///   library.  When this is `None` then no quote provider library will be
+    ///   used limiting quote generation to local quote generation only.
+    /// * `id_enclave` - The full file path to the ID enclave.
+    ///
+    /// # Errors
+    /// * [`Error::PathsInitialized`] if the paths have been previously
+    ///   initialized.
+    /// * [`Error::PathStringConversion`] if one of the paths cannot be
+    ///   converted to a [`CString`]
+    /// * [`Error::Sgx`] if
+    ///     * one of the paths does not point to a file
+    ///     * one of the paths is longer than 259 (bytes)
+    ///     * one of the paths contains a null (0) byte.
+    pub fn with_paths<P: AsRef<Path>>(
+        quoting_enclave: P,
+        provisioning_certificate_enclave: P,
+        quote_provider_library: Option<P>,
+        id_enclave: P,
```

Comment:
Consider moving these 4 fields into a holder struct
```suggestion
        paths: EnclavePaths,
```
with
```rust
struct EnclavePaths {
    quoting_enclave: PathBuf,
    provisioning_certificate_enclave: PathBuf,
    quote_provider_library: Option<PathBuf>,
    id_enclave: PathBuf,
}

Reply:
I disagree with this suggestion, but what should change about this is using `P1, P2, P3, P3`. Specifically, I'm guessing that if you tried to set paths from a mix of `String`, `&str`, `&Path`, and `PathBuf` types you'd get failures because they aren't all the same `P` type.

Reply:
>  disagree with this suggestion, but what should change about this is using P1, P2, P3, P3
To clarify @james you *disagree* with the suggestion, 
but you're saying the current implementation should be using 4 different values to allow mixing asref path types.

Curious why do you disagree? 
One advantage of the container is it can prevent transposing `with_paths(pce, qe, Some(id), qpl)`

Reply:
@jcape is right that you need four type variables in the current implementation.

I'm ambivalent about @remoun's suggestion... it is pretty common in Rust but it really only helps if you want the moral equivalent of default arguments or something, IMO. I think I'm a weak "no" on `EnclavePaths`.

Reply:
To clarify, I meant a named struct would provide a compile-time check that the paths are passed in correctly.

Reply:
Updated to use a type per path

Reply:
Yeah, the difference is really between these two calling patterns:

```rust
func(path1, path2, path3, path4);
func_s(Paths { path1, path2, path3, path4 });
```

There is an added margin of safety that's added here, in it does require that your calling code uses the same names as the function args (which it may not):

```rust
func_s(Paths {
    path1: path2,
    path2: path1,
    path3,
    path4,
});
```

Reply:
Yes, the extra typing can be trivially bypassed, but it would make it harder to *accidentally* swap the arguments, by making such bugs apparent at the call site, e.g. `foo_path: bar_path`